src/main/java/org/truffleruby/language/loader/RequireNode.java

```diff
@@ -80,34 +80,38 @@ private boolean requireWithMetrics(String feature, String expandedPathRaw, Dynam
     private boolean requireConsideringAutoload(String feature, String expandedPath, DynamicObject pathString) {
         final FeatureLoader featureLoader = getContext().getFeatureLoader();
         final List<RubyConstant> autoloadConstants = featureLoader.getAutoloadConstants(expandedPath);
-        if (autoloadConstants != null) {
-            if (getContext().getOptions().LOG_AUTOLOAD) {
-                String info = autoloadConstants
-                        .stream()
-                        .map(c -> c + " with " + c.getAutoloadConstant().getAutoloadPath())
-                        .collect(Collectors.joining(" and "));
-                RubyLanguage.LOGGER
-                        .info(() -> String.format(
-                                "%s: requiring %s which is registered as an autoload for %s",
-                                RubyContext.fileLine(getContext().getCallStack().getTopMostUserSourceSection()),
-                                feature,
-                                info));
-            }
+        try {
+            if (autoloadConstants != null) {
+                if (getContext().getOptions().LOG_AUTOLOAD) {
+                    String info = autoloadConstants
+                            .stream()
+                            .map(c -> c + " with " + c.getAutoloadConstant().getAutoloadPath())
+                            .collect(Collectors.joining(" and "));
+                    RubyLanguage.LOGGER
+                            .info(() -> String.format(
+                                    "%s: requiring %s which is registered as an autoload for %s",
+                                    RubyContext.fileLine(getContext().getCallStack().getTopMostUserSourceSection()),
+                                    feature,
+                                    info));
+                }
 
-            for (RubyConstant autoloadConstant : autoloadConstants) {
-                GetConstantNode.autoloadConstantStart(autoloadConstant);
-            }
-            try {
-                return doRequire(feature, expandedPath, pathString);
-            } finally {
                 for (RubyConstant autoloadConstant : autoloadConstants) {
-                    GetConstantNode.autoloadUndefineConstantIfStillAutoload(autoloadConstant);
-                    GetConstantNode.autoloadConstantStop(autoloadConstant);
-                    featureLoader.removeAutoload(autoloadConstant);
+                    GetConstantNode.autoloadConstantStart(autoloadConstant);
                 }
             }
-        } else {
+
             return doRequire(feature, expandedPath, pathString);
+        } finally {
+            final List<RubyConstant> releasedConstants = featureLoader.getAutoloadConstants(expandedPath);
+            if (releasedConstants != null) {
+                for (RubyConstant autoloadConstant : releasedConstants) {
+                    if (autoloadConstant.getAutoloadConstant().isAutoloading()) {
```

Comment:
I saw a lot of logs regarding autoloading - maybe we need to log this action too

Reply:
Yes, that sounds like it could be useful to debug autoload-related things.